policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+        self.changed_services = dict()
+        self.changed_pods = dict()
+        self.changed_endpoints = dict()
+
+    def run(self):
+        """
+        PolicyAgent.run() is called at program init to spawn watch threads and parse their responses
+        """
+        self.init_lists()
+        self.init_threads()
+
+        self.PodWatcher.start()
+        self.SvcWatcher.start()
+        self.NsWatcher.start()
+        self.EptsWatcher.start()
+
+        while True:
+            self.read_responses()
+
+    def init_lists(self):
+        _log.info("Initializing Resource Lists")
+        initNamespaces = _get_api_list("namespaces")
+        initServices = _get_api_list("services")
+        initPods = _get_api_list("pods")
+        initEndpoints = _get_api_list("endpoints")
+
+        for ns in initNamespaces["items"]:
+            self.process_resource(action="ADDED", kind="Namespace", target=ns)
+
+        for svc in initServices["items"]:
+            self.process_resource(action="ADDED", kind="Service", target=svc)
+
+        for pod in initPods["items"]:
+            self.process_resource(action="ADDED", kind="Pod", target=pod)
+
+        for ep in initEndpoints["items"]:
+            self.process_resource(action="ADDED", kind="Endpoints", target=ep)
+
+        self.nsRV = initEndpoints["metadata"]["resourceVersion"]
+        self.svcRV = initServices["metadata"]["resourceVersion"]
+        self.poRV = initPods["metadata"]["resourceVersion"]
+        self.epRV = initNamespaces["metadata"]["resourceVersion"]
+
+        _log.info("Service Resource Version = %s" % self.svcRV)
+        _log.info("Namespace Resource Version = %s" % self.nsRV)
+        _log.info("Pod Resource Version = %s" % self.poRV)
+        _log.info("Endpoints Resource Version = %s" % self.epRV)
+
+    def init_threads(self):
+        self.NsWatcher = Thread(target=_keep_watch,
+                                args=(self.q, "namespaces", self.nsRV))
+        self.SvcWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "services", self.svcRV))
+        self.PodWatcher = Thread(target=_keep_watch,
+                                 args=(self.q, "pods", self.poRV))
+        self.EptsWatcher = Thread(target=_keep_watch,
+                                  args=(self.q, "endpoints", self.epRV))
+
+        self.NsWatcher.daemon = True
+        self.SvcWatcher.daemon = True
+        self.PodWatcher.daemon = True
+        self.EptsWatcher.daemon = True
+
+    def read_responses(self):
+        """
+        Read Responses pulls a response off the Queue and processes it
+        If no responses remain in the queue, it will trigger a resync for all resource lists
+        """
+        try:
+            response = self.q.get_nowait()
+            r_json = json.loads(response)
+            r_type = r_json["type"]
+            r_obj = r_json["object"]
+            r_kind = r_obj["kind"]
+
+            _log.info("%s: %s" % (r_type, r_kind))
```

Comment:
I think this belongs as a debug log and should include more information.

Be wary of your logging here - this could potentially receive a lot of updates, thus cluttering the logs significantly.  Put things that might loop at debug level so they don't overwhelm.